4단계 - 자동차 경주(우승자) - 요구사항 반영

yamsroun opened the pull request.

yamsroun committed:
refactor: 코드 리뷰 반영 (2차)
- `Car`, `CarPosition`을 불변 객체로 변경
- `Car#move`에 `MovePolicy`를 파라미터로 받도록 수정(전략 패턴)
- 테스트 코드도 변경 대응

yamsroun committed:
docs: 4단계 기능 요구사항 작성

yamsroun committed:
feat: 4단계 기능 요구사항 구현 (1차)
- [자동차/경주] 각 자동차에 1~5자의 이름 부여 가능
- [입력] 자동차 이름 목록을 구분해서 입력 받음
- [출력] 경주 상태 출력 시 자동차 이름도 함께 출력
- 관련 테스트 코드 추가/수정

yamsroun committed:
feat: 4단계 기능 요구사항 구현 (2차)
- [자동차/경주] 최종 우승자 조회 기능 추가
- [출력] 최종 우승자의 자동차 이름 목록 출력
- 불필요 생성자 제거 및 필드명 수정
- 관련 테스트 코드 추가/수정

nooose commented on src/main/java/racingcar/racing/RacingEntries.java:
```diff
@@ -8,39 +8,60 @@
+        return cars;
+    }
+
+    public List<Car> getWinners() {
```
멱등성있게 구현하셨네요 👍

nooose replied:
`getWinners()`를 호출할 때 마다 
`maxPosition`과, `winners`가 초기화될텐데 멤버변수가 불필요할 것 같아요
지역변수로 써도되지 않나요?

yamsroun replied:
지역변수를 사용해도 됩니다.
처음에는 지역변수를 사용했다가, 클래스 분리 빌드-업을 해나가는 과정이었습니당~😅

`RacingEntries` 클래스
```java
public List<Car> getWinners() {
    RacingWinners racingWinners = new RacingWinners(this.cars);
    return racingWinners.getWinners();
}
```

`RacingWinners` 클래스
```java
public class RacingWinners {

    private final List<Car> cars;
    private final List<Car> winners = new ArrayList<>();
    private int maxPosition = 0;

    public RacingWinners(List<Car> cars) {
        this.cars = cars;
    }

    public List<Car> getWinners() { ... }

    private int addWinnerAndGetNewMaxPosition(Car car) { ... }
}
```

아래 글 참고해 주세용~
https://velog.io/@csh0034/Clean-Code-10%EC%9E%A5-%ED%81%B4%EB%9E%98%EC%8A%A4
<img width="816" alt="image" src="https://github.com/g-oop/goop-java-racingcar/assets/20655591/72311899-85eb-4d63-a252-39802e3cb097">

nooose replied:
역시나 이런 깊은 뜻이있었군요..

nooose commented on src/main/java/racingcar/entry/CarName.java:
```diff
@@ -0,0 +1,17 @@
+package racingcar.entry;
+
+public record CarName(String value) {
```
👍

nooose commented on src/main/java/racingcar/entry/CarPosition.java:
```diff
@@ -2,7 +2,7 @@
 
 public class CarPosition {
```
CarName은 record를 사용했는데
CarPosition은 record를 사용하지 않았네요 
이유가 있나요?

yamsroun replied:
그러게요. 왜 그랬을까요.
이 눔의 IntellJ한테 시달리다 보니..😅

nooose commented on src/main/java/racingcar/racing/RacingPreference.java:
```diff
@@ -1,7 +1,7 @@
 
 public record RacingPreference(
-    int carCount,
+    String[] carNames,
```
배열보다는 리스트가 편의측면에서 좋을 것 같습니다.

nooose commented on src/main/java/racingcar/ui/ConsoleRacingUi.java:
```diff
@@ -3,40 +3,59 @@
+import java.util.List;
 import java.util.Scanner;
 
 public class ConsoleRacingUi implements RacingUi {
```
리뷰남기려고 아무 클래스나 찍어서 올릴게요 이 클래스는 무시해주세요 😅

---
몇 가지 문제가 있어보입니다.

<img width="619" alt="image" src="https://github.com/g-oop/goop-java-racingcar/assets/36617783/ce3c8ff9-0a08-4ffa-aaf6-80fede481559">

<img width="664" alt="image" src="https://github.com/g-oop/goop-java-racingcar/assets/36617783/09acd1f7-cc54-4474-afad-de41aeca4236">

<img width="673" alt="image" src="https://github.com/g-oop/goop-java-racingcar/assets/36617783/950a9955-4458-4875-95c4-36d46e3eb7f1">


😄

nooose reviewed (changes requested):
불변천국이네요..
리뷰 남긴 것 중에 몇개만 해결하면 될 것 같습니다.
궁금해서 질문 남긴 것도있는데 답변 부탁드립니다. 🙇

yamsroun committed:
refactor: 코드 리뷰 반영 (1차)
- [입력] 자동차 이름 개수 및 공백 오류 수정
- [자동차/경주] RacingEntries 생성자 파라미터를 List 타입으로 변경 및 관련 팩터리 메소드 추가
- [자동차/경주] 이름 중복 유효성 검증 추가
- [자동차/경주] 경주 승리자 클래스 분리
- 관련 테스트 코드 추가/수정

yamsroun commented:
코드 리뷰 반영했습니다.
슬랙 메시지는 인자한 쌤인데, PR 메시지는 호랑이 쌤이네요~
덕분에 많이 배우고 있습니다.
계속 호랑이 쌤이 되어주세요~😍😍😍

nooose commented on src/test/java/racingcar/racing/RacingEntriesTest.java:
```diff
@@ -2,33 +2,64 @@
+    @Test
+    @DisplayName("자동차 이름은 중복될 수 없다.")
+    void create_duplicateCarName() {
+        List<String> carNames = List.of("yamsr", "gilbe", "nooos", "yamsr");
```
😂

nooose commented on src/main/java/racingcar/entry/CarPosition.java:
```diff
@@ -12,8 +12,8 @@ public CarPosition(int position) {
 
-    public void increase() {
-        position++;
+    public int increase() {
+        return position + 1;
     }
```
```suggestion
    public Position increase() {
        return new Position(position + 1);
    }
```

불변이니까 증가된 값을 줘도 될 것 같아요

yamsroun replied:
넹~😍

nooose commented on src/test/java/racingcar/entry/CarTest.java:
```diff
@@ -2,62 +2,57 @@
-        int nextPosition = car.currentPosition();
-
-        assertThat(nextPosition).isEqualTo(prevPosition);
+    @ParameterizedTest
+    @ValueSource(strings = {"a", "abc", "abcde"})
+    @DisplayName("자동차 이름은 1~5자만 가능하다.")
+    void allowedCarName(String carName) {
+        Car car = new Car(carName);
+        assertThat(car.getName()).isEqualTo(carName);
     }
```
CarName 클래스를 테스트해도 좋았을 것 같습니다.

yamsroun replied:
넹😍

nooose commented on src/main/java/racingcar/racing/RacingEntries.java:
```diff
@@ -8,39 +8,48 @@
-        this.cars.addAll(racingEntries.getEntries());
+    public static RacingEntries of(List<String> carNames, MovePolicy movePolicy) {
+        return new RacingEntries(carNames.stream()
+            .map(String::trim)
```
trim은 입력 관련 클래스에서 처리하고 넘겨받아도 좋았을 것 같아요

yamsroun replied:
넹🥰

nooose reviewed (approved):
왜 그런지 모르겠는데 어느 순간 제가 🐯가 되었네요..

다음은 마지막 5단계인데요
사실 1~4 단계 하시면서 5단계 요구사항들을 모두 반영하셨습니다. 
너무 깔끔해서 리뷰드릴 부분도 없고 그래서
다음 단계는 생략하고 여기서 멈춰도 될 것 같습니다.
바로 로또 해도 될 것 같네요 ㅋㅋ (귀찮아서 끝내는 거 절대! 아님)
